CHERICap.bsv

```diff
@@ -182,14 +205,65 @@ typeclass CHERICap #( type capT              // type of the CHERICap capability
   function SoftPerms getSoftPerms (capT cap);
   // Set the software permissions
   function capT setSoftPerms (capT cap, SoftPerms softperms);
-  // Get the architectural permissions
+
+  // Legalise hardware permissions
+  function Exact#(capT) legaliseHardPerms(capT cap);
+    let hp = getHardPerms(cap);
+    let oldCap = cap;
+    if (!(hp.permitLoad || hp.permitStore)) begin
+      hp.permitCap = False;
+    end
+
+    if (!(hp.permitCap && hp.permitLoad)) begin
+      hp.permitElevateLevel = False;
+      hp.permitLoadMutable = False;
+    end
+
+    if (!(hp.permitCap)) begin
+      hp.permissionStoreLevel = 0;
+    end
+
+    if (!hp.permitExecute) begin
+      hp.accessSysRegs = False;
+    end
+
+    let some = setHardPerms(cap, hp);
+
+    return Exact {exact: pack(getHardPerms(oldCap)) == pack(getHardPerms(some)), value: some};
+  endfunction
+
+  // Get whether the perms are legal
+  function Bool areLegalHardPerms(capT cap) = legaliseHardPerms(cap).exact;
```

Comment:
If it's not *too* much of a pain, I think `has` is a better prefix than `are` for these. Happy to defer if you feel strongly otherwise though. It just makes it clear from the name that you give it a cap rather than the extracted field.